### PR TITLE
cleanup release artifacts

### LINKED ISSUE #2904: Remove TESTONLY artifacts, and make istioctl work out of the box for pre-releases
Having two sets of tar balls and istioctl directories is problematic. 
More context: https://groups.google.com/forum/#!topic/istio-dev/qo13rpC2FiA

1. It is confusing and people often picks up the wrong one.
2. The artifacts that we are releasing does not work out of the box.
3. People are testing different artifacts that the one we are releasing.

Instead, can we just push pre-releases also to docker hub, and make pre-release istioctl work out of the box?

### PR DESCRIPTION
**Pull request by rkpagadala**
remove not needed istioctl and istioctl-stage directories
save the gcr and docker specific images under docker/*/ directory

fixes #2904

### CODE REVIEW

**Comment by mattdelco** [comment]
I'm not sure what the motivation of the change besides further hiding artifacts deeper in the directory tree (and not storing istioctl).  I think some of the tests try to pick up the istioctl on GCS, either directly or by looking at the ISTIOCTL_URL parameter in istio.VERSION.  So, some more changes might be needed for that to cleanup the reference in istio.VERSION and possibly with the tests that try to consume it.

The other part of the change is putting subdirectories under docker/ . I was originally thinking that only tars would be stored in here, and out of an additional sense of paranoia I had the scripts also specify *tar* as part of the wildcard search.  We don't necessary need to hold to this philosophy, but if in release/cloud_builder.sh someone specifies -c to disable docker generation then this mean docker/ won't get copied (around line 110), which means that the artifacts won't also get copied/reserved.  So, after this change opting out of docker also means the archives won't get preserved, and that doesn't seem very clean to me.  Options that come to mind is to use a separate archive/ directory, or just remove the support for the -c and -b parameters (I mostly added them for flexibility and for dev purposes to speed speed up the build or only test a specific part of the build that was changed).

**Comment by rkpagadala** [comment]
i changed the code not to depend on istioctl url, and am planning on not removing it from istio.VERSION also. PTAL

**Comment by mattdelco** [comment]
Offhand I can't find any uses of the istioctl_url that concern me (across istio, test-infra, release, and fortio repos).  I'd prefer to see all uses of the url go away, but it seems like it's still used as part of a dev doing a "make push" and trying to test against that (e.g., tests/e2e_istio_preinstalled.sh).  I suppose an option is to change the workflow to be based around istio tars that contain istioctl, rather than just istioctl itself, but that's probably more like a feature (likely one not everyone would like) than a bugfix.

Anyway, the main question/concern I have is what to do about ISTIOCTL_URL in daily/monthly release builds.  The istio-archive target is still passing -i to updateVersions.sh and will effectively specify a URL that points to a non-existent file.  Are we going to turn a blind eye to this, or should functionality be added to updateVersions.sh so that an option/parameter (e.g., a new -I, or passing a blank string to -i) causes ISTIOCTL_URL to not be output in the resulting istio.VERSION file?

**Comment by rkpagadala** [comment]
@mattdelco renoved istioctl_url, we need to whittle fix istio.VERSION and its uses... for now this is a small cleanup

**Comment by rkpagadala** [comment]
fixes #4022

**Comment by mattdelco** [comment]
One problem with squirreling PROXY_TAG away in Gopkg.toml is with the "vendor.check" target in the Makefile (which is run as part of "make init" so it's a commonly executed target/rule).  The Gopkg.lock/Gopkg.toml files are diff'ed against the ones under vendor/ to help confirm that vendor/ matches the expectations of the istio repo.  My concern is that the update bot will update the sha in istio repo's Gopkg.toml and thus make this check fail.  The choices would seem to be to make the diff smarter (unlikely option), or also update the vendor repo (which also means updating istio repo to also point at the newer commit of the vendor repo, which also doesn't seem a likely option), or store the PROXY_TAG somewhere else (e.g., in the Makefile itself).

**Comment by rkpagadala** [comment]
I changed the Makefile not to diff gopkg.toml between vendor and istio I think Gopkg.lock is good enough
Metadata in Gopkg.toml is supposed to carry information for other build tools by design.

File: Gopkg.toml [hunk 1]
```diff
@@ -22,6 +22,8 @@
 # TODO: fork the repo to checkin artifacts, or find an alternative repo that has artifacts available.
 # This repo includes only istio-generated files.
 
+[metadata]
```
Comment (ldemailly) [inline Gopkg.toml]: oh ? what's that ?

**Comment by ldemailly** [comment]
> I changed the Makefile not to diff gopkg.toml between vendor and istio I think Gopkg.lock is good enough

I don't think that's correct. I'm having issues avoid people messing up vendor (because of the submodule subtleties) and that's why I put those checks

specifically - the reason for that is ensure "dep status" stays error free

can you check if "[meta]" is ignored for the input sha ? ie can that section be changed and it doesn't change the output sha in the lock ? (try "make depend.status" and see you don't get errors)

File: Makefile [hunk 2]
```diff
@@ -386,7 +384,6 @@ ${ISTIO_OUT}/archive: istioctl-all LICENSE README.md istio.VERSION install/updat
 	cp -r tools ${ISTIO_OUT}/archive
 	install/updateVersion.sh -c "$(ISTIO_DOCKER_HUB),$(VERSION)" \
                                  -x "$(ISTIO_DOCKER_HUB),$(VERSION)" -p "$(ISTIO_DOCKER_HUB),$(VERSION)" \
-                                 -i "$(ISTIO_URL)/$(ISTIO_URL_ISTIOCTL)" \
```
Comment (ldemailly) [inline Makefile]: who/what is going to set that url then?

File: Gopkg.toml (hunk 1 above)
Comment (rkpagadala) [reply Gopkg.toml]: https://github.com/golang/dep/blob/master/docs/Gopkg.toml.md#metadata

File: Makefile (hunk 2 above)
Comment (rkpagadala) [reply Makefile]: We should remove or cut down istio.VERSION, the only place it is actually used is in vm mesh expansion and there we are using PILOT_DEB_URL and thats it. This is a partial cleanup, nobody is using URL and it is not being set anywhere.

Comment (ldemailly) [reply Makefile]: I would think we need to either remove it or keep it working ? not in between?

File: Gopkg.toml (hunk 1 above)
Comment (ldemailly) [reply Gopkg.toml]: nice, can you check the effect on inputs-digest
https://github.com/istio/istio/blob/master/Gopkg.lock#L1469

**Pull request closed without merging**